packages/eslint-plugin/src/rules/type-annotation-spacing.ts

```diff
@@ -32,6 +47,107 @@ const definition = {
   additionalProperties: false,
 };
 
+function createRules(options?: Config): WhitespaceRules {
+  const globals = {
+    ...(options?.before !== undefined ? { before: options.before } : {}),
+    ...(options?.after !== undefined ? { after: options.after } : {}),
+  };
+  const override = options?.overrides ?? {};
+  const colon = {
+    ...{ before: false, after: true },
+    ...globals,
+    ...override?.colon,
+  };
+  const arrow = {
+    ...{ before: true, after: true },
+    ...globals,
+    ...override?.arrow,
+  };
+
+  return {
+    colon: colon,
+    arrow: arrow,
+    variable: { ...colon, ...override?.variable },
+    property: { ...colon, ...override?.property },
+    parameter: { ...colon, ...override?.parameter },
+    returnType: { ...colon, ...override?.returnType },
+  };
+}
+
+function getNodeType(node?: TSESTree.Node): string {
+  return node?.type ?? '';
+}
+
+function typeIsArrowFunction(type: string): boolean {
+  return type === AST_NODE_TYPES.TSFunctionType;
+}
+
+function typeIsIdentifier(type: string): boolean {
+  return type === AST_NODE_TYPES.Identifier;
+}
+
+function typeIsProperty(type: string): boolean {
+  return [
+    AST_NODE_TYPES.ClassProperty,
+    AST_NODE_TYPES.FunctionExpression,
+    AST_NODE_TYPES.TSPropertySignature,
+    AST_NODE_TYPES.TSMethodSignature,
+  ].includes(type as AST_NODE_TYPES);
+}
```

Comment:
ditto for moving into `astUtils`.
This should also be expanded a bit based on all of the applicable types

https://github.com/typescript-eslint/typescript-eslint/blob/ce41d7de33bcb7ccf96c03ac1438304c5a49ff54/packages/typescript-estree/src/ts-estree/ts-estree.ts#L309-L316
https://github.com/typescript-eslint/typescript-eslint/blob/ce41d7de33bcb7ccf96c03ac1438304c5a49ff54/packages/typescript-estree/src/ts-estree/ts-estree.ts#L508-L513

```ts
function isClassOrTypeElement(
  node: TSESTree.Node | undefined,
): node is TSESTree.ClassElement | TSESTree.TypeElement {
  if (!node) {
    return false;
  }

  return [
    // ClassElement
    AST_NODE_TYPES.ClassProperty,
    AST_NODE_TYPES.FunctionExpression,
    AST_NODE_TYPES.MethodDefinition,
    AST_NODE_TYPES.TSAbstractClassProperty,
    AST_NODE_TYPES.TSAbstractMethodDefinition,
    AST_NODE_TYPES.TSEmptyBodyFunctionExpression,
    AST_NODE_TYPES.TSIndexSignature,
    // TypeElement
    AST_NODE_TYPES.TSCallSignatureDeclaration,
    AST_NODE_TYPES.TSConstructSignatureDeclaration,
    // AST_NODE_TYPES.TSIndexSignature,
    AST_NODE_TYPES.TSMethodSignature,
    AST_NODE_TYPES.TSPropertySignature,
  ].includes(node.type);
}
```

Reply:
Thanks for providing all of the exhaustive guard functions. This would have taken me ages.